### PR TITLE
OBSDATA-1354: Druid 25 with confluent patches

### PR DESCRIPTION
**Pull request by Ghazanfar-CFLT**
Apply Confluent patches on top of Druid 25.

Druid 25 release notes can be found [here](https://github.com/apache/druid/releases/tag/druid-25.0.0)

**Why suffix the branch wit `-rc1`?**
In case I messup, I don't have to work with IT to delete the `25.0.0-confluent` branch.
If [25.0.0-confluent-rc1](https://github.com/confluentinc/druid/tree/25.0.0-confluent-rc1) works out, I will just create a `25.0.0-confluent` branch from that one. If not, I'll give it another trying using `-rc2`

---
This section quotes @xvrl from slack,
### Some features of interest to us
1. Additional partition lag metrics for Kafka ingestion
2. Async task client for streaming ingestion (which I had already mentioned in November)
3. A fix for overlord leader-election locking up https://github.com/apache/druid/releases/tag/druid-25.0.0#25-ingestion-fixed-overlord-leader-election
4. A fix for tasks "getting stuck" and when overlord leader switches, which we have seen a few times https://github.com/apache/druid/releases/tag/druid-25.0.0#25-ingestion-streaming-tasks-resume-on-overlord-switch
5. Compaction now reduces the amount of disk usage https://github.com/apache/druid/releases/tag/druid-25.0.0#25-operations-compaction
6. An option to batch segment allocation, which can reduce ingestion lag when waiting for segment allocation https://github.com/apache/druid/releases/tag/druid-25.0.0#25-operations-batch-allocation
7. Improvements to reduce over-replication of segments during rebalancing https://github.com/apache/druid/releases/tag/druid-25.0.0#25-operations-segment-replication
8. Support for running tasks natively one k8s, which means we don't need to run middle-managers for compaction tasks anymore, and can bin-pack those workloads anywhere https://github.com/apache/druid/releases/tag/druid-25.0.0#25-highlights-kubernetes-native-tasks

### Caution
Task management and segment discovery will default to http by default https://github.com/apache/druid/releases/tag/druid-25.0.0#25-upgrade-segment-discovery

---

### Additional commits
1. https://github.com/confluentinc/druid/pull/136/commits/a9acfc3db8b56847e10cb0dadf57bc6c8224fb92 - Fix lint issue in CI/CD
The lint would fail. A similar fix has been made on druid 26. See [here](https://github.com/apache/druid/blob/druid-26.0.0/web-console/src/views/sql-data-loader-view/schema-step/schema-step.tsx#L519). I didn't cherry-pick the corresponding [commit](https://github.com/apache/druid/pull/13762) from apache/druid because it was very large and changes a lot of files.
2. https://github.com/confluentinc/druid/pull/136/commits/68fbe4f66d540c2bed54606723be69eabeb5c1ae - Fix jest and prettify checks
These changes were created by running the following commands,
```
npx prettier --write .
npm run jest -- -u
```

---
### Testing

<details>
<summary>
mvn clean package -DskipTests -T1C -X
</summary>

```
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Druid 25.0.0:
[INFO]
[INFO] Druid .............................................. SUCCESS [  9.509 s]
[INFO] druid-core ......................................... SUCCESS [ 35.928 s]
[INFO] druid-hll .......................................... SUCCESS [ 18.039 s]
[INFO] extendedset ........................................ SUCCESS [ 18.270 s]
[INFO] druid-processing ................................... SUCCESS [01:03 min]
[INFO] druid-aws-common ................................... SUCCESS [ 18.944 s]
[INFO] druid-gcp-common ................................... SUCCESS [ 18.145 s]
[INFO] druid-server ....................................... SUCCESS [ 45.045 s]
[INFO] druid-indexing-hadoop .............................. SUCCESS [02:06 min]
[INFO] druid-indexing-service ............................. SUCCESS [01:50 min]
[INFO] druid-sql .......................................... SUCCESS [01:15 min]
[INFO] druid-services ..................................... SUCCESS [ 29.801 s]
[INFO] druid-orc-extensions ............................... SUCCESS [01:36 min]
[INFO] druid-avro-extensions .............................. SUCCESS [01:48 min]
[INFO] druid-parquet-extensions ........................... SUCCESS [01:37 min]
[INFO] druid-protobuf-extensions .......................... SUCCESS [ 15.048 s]
[INFO] druid-s3-extensions ................................ SUCCESS [ 37.391 s]
[INFO] druid-kinesis-indexing-service ..................... SUCCESS [ 14.989 s]
[INFO] druid-azure-extensions ............................. SUCCESS [ 35.892 s]
[INFO] druid-google-extensions ............................ SUCCESS [ 35.632 s]
[INFO] druid-hdfs-storage ................................. SUCCESS [01:36 min]
[INFO] druid-datasketches ................................. SUCCESS [ 31.345 s]
[INFO] druid-histogram .................................... SUCCESS [ 27.697 s]
[INFO] mysql-metadata-storage ............................. SUCCESS [ 36.272 s]
[INFO] druid-kafka-indexing-service ....................... SUCCESS [ 15.099 s]
[INFO] druid-basic-security ............................... SUCCESS [ 38.704 s]
[INFO] druid-lookups-cached-global ........................ SUCCESS [ 38.533 s]
[INFO] druid-testing-tools ................................ SUCCESS [ 23.734 s]
[INFO] simple-client-sslcontext ........................... SUCCESS [ 35.615 s]
[INFO] druid-multi-stage-query ............................ SUCCESS [ 33.586 s]
[INFO] druid-integration-tests ............................ SUCCESS [ 15.870 s]
[INFO] druid-stats ........................................ SUCCESS [ 26.579 s]
[INFO] druid-benchmarks ................................... SUCCESS [16:36 min]
[INFO] web-console ........................................ SUCCESS [03:13 min]
[INFO] druid-kubernetes-extensions ........................ SUCCESS [ 40.437 s]
[INFO] druid-bloom-filter ................................. SUCCESS [ 26.514 s]
[INFO] druid-kerberos ..................................... SUCCESS [ 37.684 s]
[INFO] druid-pac4j ........................................ SUCCESS [01:32 min]
[INFO] druid-kafka-extraction-namespace ................... SUCCESS [ 27.751 s]
[INFO] postgresql-metadata-storage ........................ SUCCESS [ 36.529 s]
[INFO] druid-lookups-cached-single ........................ SUCCESS [ 37.795 s]
[INFO] druid-ec2-extensions ............................... SUCCESS [ 12.903 s]
[INFO] druid-aws-rds-extensions ........................... SUCCESS [ 18.024 s]
[INFO] druid-ranger-security .............................. SUCCESS [ 27.928 s]
[INFO] druid-catalog ...................................... SUCCESS [ 27.944 s]
[INFO] druid-compressed-bigdecimal ........................ SUCCESS [ 26.610 s]
[INFO] druid-influx-extensions ............................ SUCCESS [ 18.217 s]
[INFO] druid-cassandra-storage ............................ SUCCESS [ 18.025 s]
[INFO] dropwizard-emitter ................................. SUCCESS [ 23.805 s]
[INFO] druid-cloudfiles-extensions ........................ SUCCESS [ 17.012 s]
[INFO] graphite-emitter ................................... SUCCESS [ 19.587 s]
[INFO] druid-distinctcount ................................ SUCCESS [ 10.604 s]
[INFO] statsd-emitter ..................................... SUCCESS [ 17.709 s]
[INFO] druid-time-min-max ................................. SUCCESS [ 18.464 s]
[INFO] druid-virtual-columns .............................. SUCCESS [ 10.741 s]
[INFO] druid-thrift-extensions ............................ SUCCESS [01:37 min]
[INFO] ambari-metrics-emitter ............................. SUCCESS [ 18.516 s]
[INFO] sqlserver-metadata-storage ......................... SUCCESS [ 19.559 s]
[INFO] kafka-emitter ...................................... SUCCESS [ 24.949 s]
[INFO] druid-redis-cache .................................. SUCCESS [ 26.273 s]
[INFO] druid-opentsdb-emitter ............................. SUCCESS [ 18.605 s]
[INFO] materialized-view-maintenance ...................... SUCCESS [ 12.990 s]
[INFO] materialized-view-selection ........................ SUCCESS [ 10.710 s]
[INFO] druid-momentsketch ................................. SUCCESS [ 16.658 s]
[INFO] druid-moving-average-query ......................... SUCCESS [ 24.724 s]
[INFO] tdigestsketch ...................................... SUCCESS [ 24.468 s]
[INFO] druid-influxdb-emitter ............................. SUCCESS [ 10.831 s]
[INFO] gce-extensions ..................................... SUCCESS [ 12.939 s]
[INFO] aliyun-oss-extensions .............................. SUCCESS [06:06 min]
[INFO] prometheus-emitter ................................. SUCCESS [ 24.670 s]
[INFO] opentelemetry-emitter .............................. SUCCESS [ 27.024 s]
[INFO] druid-kubernetes-overlord-extensions ............... SUCCESS [ 15.149 s]
[INFO] druid-opentelemetry-extensions ..................... SUCCESS [ 24.180 s]
[INFO] druid-opencensus-extensions ........................ SUCCESS [ 46.575 s]
[INFO] confluent-extensions ............................... SUCCESS [ 10.419 s]
[INFO] distribution ....................................... SUCCESS [ 28.154 s]
[INFO] druid-it-tools ..................................... SUCCESS [ 27.928 s]
[INFO] druid-it-image ..................................... SUCCESS [ 11.596 s]
[INFO] druid-it-cases ..................................... SUCCESS [ 12.904 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  21:15 min (Wall Clock)
[INFO] Finished at: 2023-05-30T11:38:07+05:30
[INFO] ------------------------------------------------------------------------
```
</details>

### CODE REVIEW

File: distribution/docker/Dockerfile [hunk 1]
```diff
@@ -18,36 +18,15 @@
...
 
 ARG JDK_VERSION=11
+FROM alpine as extractor
 
-# The platform is explicitly specified as x64 to build the Druid distribution.
-# This is because it's not able to build the distribution on arm64 due to dependency problem of web-console. See: https://github.com/apache/druid/issues/13012
-# Since only java jars are shipped in the final image, it's OK to build the distribution on x64.
-# Once the web-console dependency problem is resolved, we can remove the --platform directive.
-FROM --platform=linux/amd64 maven:3.8.6-jdk-11-slim as builder
+ARG VERSION=
```
Comment (saithal-confluent) [inline distribution/docker/Dockerfile:23]: Is this intentionally set to blank?

File: web-console/README.md
Comment (saithal-confluent) [inline web-console/README.md:1]: Should we revert the changes in this file to keep it closer to upstream and hence limit changes in number of files while doing an upgrade?

File: web-console/src/ace-modes/hjson.js
Comment (saithal-confluent) [inline web-console/src/ace-modes/hjson.js:1]: Should we revert the changes in this file to keep it closer to upstream and hence limit changes in number of files while doing an upgrade?

File: .semaphore/semaphore.yml [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        - name: "Kafka Indexing Service"
+          env_vars:
+            - name: MAVEN_PROJECTS
+              value: extensions-core/kafka-indexing-service
+          commands: *run_tests
+
+        - name: "Other Tests"
+          env_vars:
+            - name: MAVEN_PROJECTS
+              value: '!server,!processing,!indexing-service,!extensions-core/kafka-indexing-service,!extensions-contrib/confluent-extensions,!integration-tests-ex/cases'
```
Comment (saithal-confluent) [inline .semaphore/semaphore.yml:132]: There's a new extension added for running kubernetes tasks as jobs, should we add that here or is it taken care somewhere else?

**Review by saithal-confluent — COMMENTED** [review]
Left a few comments. Please check

File: distribution/docker/Dockerfile (hunk 1 above)
Comment (Ghazanfar-CFLT) [reply distribution/docker/Dockerfile:23]: Yeah, that's set during build time. It gets set in [distribution/pom.xml](https://github.com/confluentinc/druid/pull/136/files#diff-ac19e9d7369629bf473b9b6c514012d5fecc319f98a6b9b57855f3202e970610R795)

As a part of this commit https://github.com/confluentinc/druid/pull/136/commits/d4648d22bda94760e5f6b18f049c17e93379fabe

File: web-console/README.md
Comment (Ghazanfar-CFLT) [reply web-console/README.md:1]: @saithal-confluent without this the build would fail - I had to made this change to get the build working. It was an automated change - see PR description for the exact commands that made this change.

File: web-console/src/ace-modes/hjson.js
Comment (Ghazanfar-CFLT) [reply web-console/src/ace-modes/hjson.js:1]: @saithal-confluent without this the build would fail - I had to made this change to get the build working. It was an automated change - see PR description for the exact commands that made this change.

File: .semaphore/semaphore.yml (hunk 2 above)
Comment (Ghazanfar-CFLT) [reply .semaphore/semaphore.yml:132]: @saithal-confluent won't that automatically be included here? To me it looks like we are excluding projects from here which already have their own dedicated job. Right?

File: web-console/src/ace-modes/hjson.js
Comment (saithal-confluent) [reply web-console/src/ace-modes/hjson.js:1]: I missed that. Thanks.

**Review by saithal-confluent — APPROVED** [review]

**Review by saithal-confluent — APPROVED** [review]

File: distribution/docker/Dockerfile [hunk 3]
```diff
@@ -18,36 +18,15 @@
...
-      ; fi
+COPY ./target/apache-druid-${VERSION}-bin.tar.gz .
 
-RUN --mount=type=cache,target=/root/.m2 VERSION=$(mvn -B -q org.apache.maven.plugins:maven-help-plugin:3.2.0:evaluate \
-      -Dexpression=project.version -DforceStdout=true \
-    ) \
- && tar -zxf ./distribution/target/apache-druid-${VERSION}-bin.tar.gz -C /opt \
- && mv /opt/apache-druid-${VERSION} /opt/druid
+RUN tar -zxf /src/apache-druid-${VERSION}-bin.tar.gz -C /opt \
+ && ln -s /opt/apache-druid-${VERSION} /opt/druid
```
Comment (xvrl) [inline distribution/docker/Dockerfile:29]: there were quite a few changes upstream to the docker build (e.g. https://github.com/apache/druid/pull/11299) If you've tested the build and it work then we can do it as a follow-up but, since it looks like this was done prior to 24.x and we missed it, but this might be a good time to do it and avoid further drift since it might bite us later.

Comment (Ghazanfar-CFLT) [reply distribution/docker/Dockerfile:29]: Hey @xvrl !
I have tested the build - I will submit a follow up PR for this.
